zilliqa/tests/it/consensus.rs

```diff
@@ -1,11 +1,138 @@
 use crate::CombinedJson;
+use crate::Network;
 use ethabi::Token;
-
+use ethers::{providers::Middleware, types::TransactionRequest};
+use tracing::*;
 use zilliqa::state::Address;
 
-use crate::Network;
+// Test that a node that joins later can sync up to the latest block and become a
 
-use ethers::{providers::Middleware, types::TransactionRequest};
+/*
+// Test that a node can die and rejoin and still sync up to the latest block
+#[zilliqa_macros::test]
+async fn node_can_sync_join_block_production(mut network: Network) {
+    let start_block = 5;
+
+    // wait until at least 5 blocks have been produced
+    network
+        .run_until(
+            |n| {
+                let index = n.random_index();
+                n.get_node(index)
+                    .get_finalized_height()
+                    >= start_block
+            },
+            50,
+        )
+        .await
+        .unwrap();
+
+    // Add a new node, then make sure they eventually are the producer of
+    // a block
+    info!("Adding node to network");
+    let index = network.add_node();
+
+    network
+        .run_until(
+            |n| {
+                let block = n.node_at(index)
+                        .get_latest_block()
+                        .unwrap()
+                        .unwrap();
+
+                    block.committee.leader_index(block.number()) == index
+            },
+            50000,
+        )
+        .await
+        .expect("expected to find our joined node has become a block producer");
+}
+*/
+
+//fn network_can_die_restart() {}
+
+// Test that all nodes can die and the network can restart (even if they startup at different
+// times)
+#[zilliqa_macros::test]
+async fn network_can_die_restart(mut network: Network) {
+    let start_block = 5;
+    let finish_block = 10;
+
+    let samples: usize = std::env::var_os("ZQ_TEST_SAMPLES")
+        .map(|s| s.to_str().unwrap_or("1").parse().unwrap_or(1))
+        .unwrap_or(1);
+
+    println!("{}", samples);
+
+    // wait until at least 5 blocks have been produced
+    network
+        .run_until(
+            |n| {
+                let index = n.random_index();
+                n.get_node(index).get_finalized_height() >= start_block
+            },
+            50,
+        )
+        .await
+        .unwrap();
+
+    // Forcibly restart the network, with a random time delay between each node
+    network.restart();
+
+    // Panic if it can't progress to the target block
+    network
+        .run_until(
+            |n| {
+                let index = n.random_index();
+                n.get_node(index).get_finalized_height() >= finish_block
+            },
+            50000,
```

Comment:
This surely could use a more reasonable timeout for network startup - maybe 1000 or even 500?

Reply:
I will set it lower, but the test exits when the condition is true or at tick amount, so having the amount be high doesn't increase the test time or anything (except when it fails).